Add a Node.js-like module loader extra

fatcerberus opened the pull request:
A module loading framework and `require()` implementation based on Node.js.  Semantics match Node.js wherever possible and most features of the Node module system are provided.

https://nodejs.org/api/modules.html

Features:
* Application provides callbacks to perform module resolution and loading
* Module caching based on resolved module ID
* Tolerates circular `require()` chains
* Supports `module.exports`, `module.loaded`, and `module.require`
* Supports `require.cache`
* Supports `__filename` (but not `__dirname`)

Checklist:
- [x] Expose `duk__push_require_function()` to calling code? => `duk_module_node_init()` registers a global `require`, so not needed
- [x] Manual tests to ensure correctness of all features such as `module.loaded`, `require.cache`, etc.
- [x] Full documentation, in particular for callbacks
- [x] More/better sanity checks/guards around callback invocations? => callbacks are now Duktape/C functions
- [x] Rename extra to `module-node`
- [x] Match Node.js module ID semantics (`module.id === module.filename`)

fatcerberus commented:
One thing that bugs me about this right now is the resolve_cb: It takes the target module ID and originating module ID on the stack (not so bad), but is expected to push the filename and resolved module ID in reply.  Returning two values from a callback this way seems awkward.  Maybe that's just me though.

load_cb is more straightforward: It takes the `module` object on the stack and is expected to push the source code in response (or return 0 if it loads `module.exports` itself)

svaarala commented:
The extra could maybe be renamed `module-xxx` as it's likely there will be several loader examples.

One of the loaders should match Duktape 1.x as closely as possible IMO (same is true for the other extras to ease migration). Another that would be useful would support DLL loading for Linux, Windows, OS X.

fatcerberus commented:
That's fine, so I can rename this one to `module-node` since it's meant to emulate Node.js semantics.  In that case I could probably remove the filename/module ID distinction since in Node.js `module.id === module.filename`.

Agree on a Duktape 1.x compatible loader, that's something for a separate pull. :)

fatcerberus commented:
Let me know if there are any problems with the code.  There might be sanity checks missing, I get overwhelmed after a while shuffling the value stack around. :)

saghul commented:
This is awesome, thanks for sharing!

fatcerberus commented:
@saghul No problem, I get a lot of mileage out of Duktape so I like to give back when I can. :)

svaarala commented:
`module-node` sounds good to me if it's indeed replicating Node.js semantics - something I think many users will find useful.

The main thing w.r.t. merging is that I'm not yet sure what form the 2.x module changes will concretely take, I'll need to think about the ES6 stuff a bit first. On the other hand, whatever changes will then be needed should be relatively minor, so maybe it's better to merge and look at these changes later.

A smaller consideration is the question: what should be distributed as extras? In general, stuff that can be added to an application can be part of the Duktape distributable or simply come from an external repo. Some thoughts on that:

- The extra is part of the release and is maintained with the release (and the release cycle: extras should be fixed for a new release etc, unless they are no longer applicable).
- They should also, ideally, have some form of version compatibility which tracks Duktape version semantics at least in most cases (otherwise they're not really part of the release cycle). In other words, if you're using `extras/xyz` and you upgrade Duktape to a new minor version, there should be a high probability the updated extra would work (source compatibility).
- The functionality provided is quite close to the core Duktape functionality. For example, it might not make sense to bundle something like DukLuv with the release because it goes way beyond core functionality. Module loading, on the other hand, is quite close to the core functionality.
- Right now extras have very sparse testing coverage, but my intention is that both examples/ and extras/ would have test coverage for CI.

Not really 100% clear guidelines, and there's also a distinction between examples/ and extras/ (examples not being "production quality") which is somewhat artificial. Doing changes here before 2.0 release may be useful. For example, maybe all extras could come from external repos and simply be included in the dist phase? That way the issue tracking related to them would be kept separate.

fatcerberus commented:
Yeah, that's my intent--to replicate Node semantics as far as possible.  Doing so requires two callbacks because Node.js maps concrete file names to module IDs, so it's up to the application to deal with the file system.

I agree re: versioning.  The readme for extras says that semantic versioning doesn't apply to them, but it's still nice to try to maintain compatibility where possible.  In practice that shouldn't be too difficult since Duktape itself is source-compatible between minor releases.

Note that this does still need some cleanup before you go merging it. :)

svaarala commented:
@fatcerberus Right, the extras don't strictly follow the versioning because that might be pretty difficult for release management. But even so, there should IMO be a pretty high expectation they're changed with versioning in mind. Otherwise it's somewhat confusing that they're part of the versioned release cycle IMO.

svaarala commented:
Added a separate issue for the extras/examples stuff: https://github.com/svaarala/duktape/issues/797.

fatcerberus commented:
I totally agree.  Like I said Duktape being source-compatible between minor versions helps a lot here. :)

fatcerberus commented:
Getting back to this pull in particular: The callbacks seem fragile.  They are called directly like the debugger callbacks and can do evil things to the value stack, such as popping off the `module` object under construction.

Would it be worthwhile to add sanity checks after the callback invocations to verify stack shape, or leave it as-is under the assumption that the callbacks are trusted since they are C functions?

fatcerberus commented:
I added a basic test harness.  Currently it tests for:

* Whether the ID arrived at by resolve_cb makes it into load_cb
* Proper caching behavior
* Correct behavior of `require()` from both global code and inside a module

Ideally the test should be more comprehensive, but I'm happy to report the basic tests described above pass. :-)

fatcerberus commented:
Thinking about the callbacks some more, it might be nice if `duk_safe_call()` gave the called function its own stack frame.  This would be useful to users as well as internally for callbacks: For example the debugger could call the app request callback isolating it from the rest of the value stack.

fatcerberus commented:
Alright, I improved the tests.  Coverage should be pretty comprehensive now.  This is merge ready for my part, pending code review of course.

svaarala commented:
@fatcerberus Cool, I'll take a look at this as soon as possible. One thing from a quick glance is the file/function naming: it would be good if the file and function names matched the directory name, i.e. duk_module_node.c, duk_module_node_init(), etc. For example duk_module_init() is rather generic and it's not apparent from an init call site which loader variant is being initialized.

fatcerberus commented:
Hm, looking over the Node Module API again, I could probably also implement support for `module.children` and `module.parent` easily enough, I'll look into that.

fatcerberus commented:
I decided not to implement `module.children` and `module.parent` as doing so is not trivial (`require()` tracks the parent ID, but not the `module` object) and these seem like pretty niche features.  The feature set implemented here is plenty for a Node-like loader, I think.

I did go ahead and change the init function's name to `duk_module_node_init()` though.

svaarala commented on extras/module-node/README.rst:
```diff
@@ -0,0 +1,54 @@
+          const char *parent_id = duk_get_string(ctx, -2);  /* calling module ID */
+          const char *module_id = duk_get_string(ctx, -1);
+
+          /* Arrive at the canonical module ID in whatever way makes sense */
```
Trivia: comments starting with capitals should end in a period (sentences).

svaarala commented on extras/module-node/README.rst:
```diff
@@ -0,0 +1,54 @@
+      duk_ret_t cb_load_module(duk_context *ctx) {
+          /* `module` object on stack top, `module.id` is the resolved ID */
+
+          /* Retrieve the module's JS source code */
```
Trivia: period.

svaarala commented on extras/module-node/README.rst:
```diff
@@ -0,0 +1,54 @@
+          /* Retrieve the module's JS source code */
+
+          duk_push_string(ctx, source_string);
+          return 1; /*nrets*/
```
Trivia: typically there are two spaces between a semicolon and a same-line comment (just for some visual separation).

svaarala commented on extras/module-node/test.c:
```diff
@@ -0,0 +1,68 @@
+}
+
+static duk_ret_t cb_load_module(duk_context *ctx) {
+	const char* filename;
```
Trivia: the star placement is `const char *xxx` in Duktape sources.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+ *
+ * https://nodejs.org/api/modules.html
+ */
+
```
Trivia: there are two basic comment styles in the Duktape code base. First style is "compact":

    /* Foo bar.
     * Quux.
     */

In this style there's a single space between the star and the text. The second style is a "banner":

    /*
     *  Two spaces.
     */

Here there are two spaces before the text. This is pretty arbitrary but that's the current convention in any case. The basic intent is to visualize separate the banner as a block separate, and the compact comment as applying to the next function, variable, etc (and there's usually no empty line between the compact comment and the function/variable).

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+
+static duk_bool_t duk__get_cached_module(duk_context *ctx, const char *id) {
+	duk_push_global_stash(ctx);
+	duk_get_prop_string(ctx, -1, "\xffModuleCache");
```
The internal key here (and in other places) follows the format reserved for Duktape internals. Extras are maybe a bit of a grey area in this respect, so I'm not sure what'd be best.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+		duk_remove(ctx, -2);
+		return 1;
+	}
+	else {
```
Trivia: else-style:

    } else {

fatcerberus replied:
Whoops, I should know better than to edit Dujtape code in MSVC, it likes to do this automatically.  I like the style but it doesn't match Duktape conventions. :)

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+ * NOTE: duk__put_cached_module() leaves the module object on the
+ * stack as a convenience to the caller.
+ */
+static void duk__put_cached_module(duk_context *ctx) {
```
Trivia: block comment style (two spaces).

fatcerberus replied:
Or maybe it should be a compact comment instead?

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+ */
+static void duk__put_cached_module(duk_context *ctx) {
+	/* ... module */
+
```
Trivia (very): in representing value stack shape I've usually used brackets to make the top and bottom explicit:

    [ foo bar quux]

and when the bottom is not known or isn't interesting:

    [ ... bar quux ]

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+	if (nrets > 1) {
+		/* Pop off extraneous return values. */
+		duk_pop_n(ctx, nrets - 1);
+	}
```
This seems a bit awkward (I guess it's related to the safe call issue). Couldn't the resolve function simply be trusted to leave the stack in a certain shape (i.e. never return > 1). Or maybe throw an error for `nrets > 1`?

The popping behavior seems a bit odd because it doesn't actually match Duktape/C behavior because the topmost value is discarded.

fatcerberus replied:
The idea was if the function claims to return multiple values we only want the first one pushed, but I agree the behavior is odd.  I guess I could change it to be a `void`-returning function, but for me personally the `duk_ret_t` return is a convenient indicator that Duktape expects something on the value stack.

I'll change it to throw an error if `nrets != 1` I guess.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+	duk_idx_t module_idx;
+
+	duk_push_current_function(ctx);
+	duk_get_prop_string(ctx, -1, "\xffModuleID");
```
Cast to void to signal intentional return value ignore:

     (void) duk_get_prop_string(...)

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+	duk_push_current_function(ctx);
+	duk_get_prop_string(ctx, -1, "\xffModuleID");
+	parent_id = duk_get_string(ctx, -1);
+
```
Since this is not performance critical code, maybe use `duk_require_string()` here. It will cheaply ensure `parent_id` will be non-NULL without any further checks.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+	id = duk_require_string(ctx, 0);
+
+	duk_push_string(ctx, parent_id);
+	duk_push_string(ctx, id);
```
Here maybe just use `duk_dup()`? Using `duk_push_string()` causes an unnecessary re-interning of the string already on the value stack?

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+	if (duk__get_cached_module(ctx, id)) {
+		goto have_module;  /* use cached module */
+	}
+	else {
```
Trivia: else style.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+			/* Pop off extraneous return values. */
+			duk_pop_n(ctx, nrets - 1);
+		}
+		if (nrets < 0 || (nrets > 0 && !duk_is_string(ctx, -1))) {
```
Same nrets issue as above.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+			duk_pop_n(ctx, nrets - 1);
+		}
+		if (nrets < 0 || (nrets > 0 && !duk_is_string(ctx, -1))) {
+			duk_error(ctx, DUK_ERR_API_ERROR, "error in module loading");
```
Should there be a duk__del_cached_module() here too?

fatcerberus replied:
Yes, good catch.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+	 * to implement CommonJS semantics and matches the behavior of Node.js.
+	 * The wrapper has no line breaks so it won't interfere with line
+	 * numbering.
+	 */
```
Trivia: block comment style.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+
+	return DUK_EXEC_SUCCESS;
+
+fail:
```
Label indent by one space.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+static void duk__push_module_object (duk_context *ctx, const char *id);
+
+static duk_resolve_module_function resolve_module_cb;
+static duk_load_module_function load_module_cb;
```
Having these as file locals means that different resolve/load functions cannot be used for different Duktape global environments or different Duktape heaps?

saghul replied:
Side question: is there a way to attach stuff to a Duktape heap? Like some user data `void*` for instance.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,248 @@
+	/* global `require` is not enumerable */
+	duk_def_prop(ctx, -3, DUK_DEFPROP_HAVE_VALUE
+		| DUK_DEFPROP_SET_WRITABLE
+		| DUK_DEFPROP_SET_CONFIGURABLE);
```
The indent style makes the flags a bit difficult to read. The typical style for this would be to tab indent for the base level (`duk_def_prop()`) and then space indent to align the rest:

    duk_def_prop(ctx, -2, DUK_DEFPROP_HAVE_VALUE |
                          DUK_DEFPROP_SET_WRITABLE |
                          DUK_DEFPROP_SET_CONFIGURABLE);

svaarala commented:
@fatcerberus Ok, some high level questions / comments from review:

- Out of curiosity, how does the resolve/load process here compare to #60 and DukLuv's loader mechanism? While it's not critical for extras to align, it'd also be good to avoid differences where they don't need to exist.
- The resolve/load callbacks are now called directly without protection. Could they be called via a combination of `duk_push_c_function()` + `duk_call()` instead? They only return (at most) one return value so that should be possible I think? (This is also related to the separate discussion on changing safe call semantics.)
- Overall using custom non-Duktape calls for resolve/load seems awkward. If they remain as is, maybe a short comment on the required call stack behavior for the resolve/load callbacks.
- The loader currently stores resolve/load function references to a file local variable which won't work with multiple different loader configurations (in separate global objects or Duktape heaps). On the other hand, stashing a custom function pointer is not supported by the Duktape API now (that's actually something to maybe fix as it's OK to cast between function pointer types, but function pointer size may differ from data pointer size, #808).
- The resolve/load callbacks cannot be Ecmascript functions at present. This would be fixed if they were provided as function references which could be replaced. But I'm guessing it's intentional they must be C functions? For me it'd be more natural if they were Duktape/C functions (C or Ecmascript) and stored in the stash to isolate them from Ecmascript code.
- What's the intended behavior when used from master as it is now, i.e. with `require()` still part of the built-ins? I'd guess it'd replace the built-in module loader? This probably doesn't warrant mentioning however, as it's likely the built-in require() will disappear altogether.

Formatting trivia etc in commit comments.

svaarala commented:
So actually most of the high level comments are about the resolve/load functions. If they were represented as Duktape/C functions and stashed as bona fide function references in the stash, several problems would be solved at once. Would there be some concrete issue in doing so?

fatcerberus commented:
The callbacks ended up as pure C calls because, in my first draft, I was passing and returning values directly.  I ended up changing it to just use the valstack because that made it easier to deal with strings which would otherwise need to be dynamically allocated (awkward) or stored in a static buffer (also awkward, and a potential security issue).

And yes, the static pointers were because I couldn't figure out a safe way to stash them.  As you point out, it's not safe to cast function pointers to `void*` (not if you care about portability anyway :).

It should be possible to change them to be Duktape/C calls now, since their signatures match anyway.  I agree that's the best solution here.  It would also solve the issue of dealing with extra garbage values pushed by the callbacks, since Duktape call handling would clean that up automatically.

The `require` registered in `duk_module_node_init()` _should_ override Duktape's, assuming that's registered at heap creation.  If not (e.g. if it's actually registered on setting `Duktape.modSearch`) then there might be an issue.  But again, like you say, it might just be better to move `require` out of the core entirely.  It isn't part of Ecmascript spec, and ES6 modules will need a different mechanism anyway.

Regarding overlap with #60: The resolve/load mechanism was indeed inspired by that implementation, however this also provides several features from Node (like `module.loaded`) which aren't supported by the minimal system built into Duktape, so I think it justifies its existence.  One change that might be worth making (perhaps in a future pull) is to expose the callbacks to Ecmascript, i.e. `Duktape.modResolve` and `.modResolve`.

Anyway, I'll go through the code and fix the trivia first, then we'll see about maybe improving the semantics before the merge.

fatcerberus commented:
One issue with `module.loaded` handling is that if a module does `Object.freeze(module)` then the loader will throw an error when trying to set `.loaded` to true.  Node also throws an error in this circumstance though, so I think it's okay.  Maybe consider it a failed load and delete the module from the cache before rethrowing?

Anyway I did the mentioned fixes and made some enhancements, but I have no internet access right now due to a Comcast outage, so I'll push the changes when I can.  For now I need sleep. :)

svaarala commented:
Re: #60, one important question is whether this pull addresses/implements the proposed design. It doesn't have to, but I'd like to understand what else is needed. (I'll go through the module loading issues as soon as I can; there are many pending issues and ideas to address; an extra with DLL loading support is one.)

fatcerberus commented:
The proposed design was to split `Duktape.modSearch` into modResolve and modLoad, in which case yes, this extra implements that--with the caveat that the callbacks are set from C once on initialization and are inaccessible from Ecmascript.

Honestly I think that makes more sense - if you're going to implement your module loader in JS anyway you might as well go the whole hog and replace `require` too.

fatcerberus commented:
One thing this doesn't have that the Dukluv implementation does is a concept of a "compile" callback, i.e. `Duktape.modCompile`.  I didn't implement that here, mainly because I couldn't figure out what problem it was supposed to solve. :)

svaarala commented:
@fatcerberus There are many cases where it makes quite a lot of sense for modResolve / modLoad to be implemented in Ecmascript (calling into C for low level operations like looking for DLLs, or loading stuff over the network), so being able to do that would be nice. Some applications are written with the style where all logic and state handling is in Ecmascript side and C calls are fully stateless helpers only, and don't handle stuff like search paths, etc.

However, being able to register the callbacks from Ecmascript code is a different issue, and I agree it might be best if that wasn't possible by default. For example, the handlers could be pushed on the value stack before calling `duk_module_node_init()` and then be registered into the global stash for later lookup. Ecmascript code couldn't access them directly, but they could still be implemented in Ecmascript.

fatcerberus commented:
Oh, I don't doubt the utility of implementing the callbacks in ES--I just meant that if you're already doing that, it's probably cleaner at that point just to replace `require()`.

fatcerberus commented:
@svaarala I pushed the fixes.

svaarala commented:
Sure you could write another provider. But what if you wanted Node.js module semantics?

It seems counterproductive for me if Duktape extras don't support writing stuff in Ecmascript :-)

svaarala commented:
As a general design point, callbacks/providers are represented as Duktape/C functions wherever possible: they have existing call mechanics, and can be implemented as Duktape/C functions, lightfuncs, or Ecmascript functions. The exceptions to this basic principle are callbacks which are closely tied to internals so that the side effects of call handling are not desirable - for example, debugger transport callbacks.

fatcerberus commented:
@svaarala Aw, crud.  I didn't have internet access so I couldn't get the new code into my Ubuntu VM.  It seems the current code is broken.  I'll fix it and re-push.

svaarala commented:
By the way, it might seem registering callbacks as arbitrary Duktape functions seems heavy but that's actually not the case. If you store a lightfunc property in the global stash for example, the cost of that is 25 bytes on a 64-bit platform and 13 bytes on a 32-bit platform. There's no Ecmascript object to represent the lightfunc; all the lightfunc information needed for a call are in the `duk_tval`. The cost is actually the same as storing a plain pointer.

So, unless the call mechanics themselves are undesirable, it's hard to justify not using a normal function call :-)

fatcerberus commented:
Well, it isn't so much the footprint that bothers me, but the fact that a callback which might be confined to C for sandboxing reasons can nonetheless show up in an Ecmascript stack trace.

fatcerberus commented:
Okay, I fixed it.  All the tests are okay, except for `require.cache`.  That test is failing, and I can't figure out why.  To elaborate, the full test is:

* `require()` a module
* Check whether it's visible in `require.cache`
* `delete require.cache[id];`
* `require()` the module again
* Check whether `mod1 !== mod2`

For some reason the last check is failing and I have no idea what's causing it.  The load callback is clearly being called twice.  Maybe you can figure it out.

svaarala commented:
Hmm, what's the harm of showing up in a stack trace? Are you worried about the .fileName/.lineNumber blaming or just in general C filenames showing up in stack traces? The blaming behavior can be fixed for 2.x.

svaarala commented:
Regarding the call style etc, I'm not actively trying to be difficult by the way :) I'm trying to figure out what kind of stuff is useful to be distributed as extras, and one criterion for that (for me) is that the extras are as generally useful as possible and play well with concepts native to Duktape. That doesn't mean other kinds of modules wouldn't be useful, just that they might not be ideal to be distributed with Duktape. Providing wiki links to useful stuff like module loaders, logging frameworks, etc, is another way to promote useful modules.

fatcerberus commented:
> Hmm, what's the harm of showing up in a stack trace?

To be honest, mostly just OCD on my part.  It just seems odd to have a function show up in the ES stack that isn't actually visible to Ecmascript in any way.  Of course if we change it to make the callbacks accessible to Ecmascript then that's a moot point. :)  That said, I did convert the callbacks to Duktape/C calls since it makes for more sane value stack semantics (the functions can use positive indices to access parameters).

The blaming is potentially an issue though - ideally an error thrown by one of the callbacks (assuming they are only accessible from C) would be blamed on the function calling `require()`.

Regarding inclusion: I think it does make sense to have an example module loader as part of the Duktape dist, especially if built-in `require()` is going to be axed (in which case we'll need a Duktape 1.x polyfill as well).  And since Node.js is the most well-known module system, it's worthwhile to emulate its semantics I think.

svaarala commented:
There's going to be at least a Duktape 1.x compatible module loader in the extras for migration (it will be minimal and match Duktape 1.x conventions, e.g. using `Duktape.modLoaded[]` etc). And there's been a lot of requests for a default loader supporting a search path and DLL loading (for DLLs which match the recommended "module conventions") for major platforms. DLL support would provide an easy default for module sharing; it's possible now but there's somewhat of a barrier without a DLL-capable loader in the distributable.

fatcerberus commented:
I can implement the Duktape 1.x loader (in a separate pull), but I'd need your help to make sure all the semantics are correct. :)

svaarala commented:
I was more or less going to just copy the existing `require()` implementation from duk_bi_global.c into an extras directory, and add a `DUK_CMDLINE_MODULE_SUPPORT` define to build it into "duk". The existing testcases would then have pretty good coverage for the loader.

fatcerberus commented:
Anyway, I rebased and squashed, waiting on the next round of review.

P.S. There have been a ton of `-Wshadow` warnings showing up in recent Travis runs.  Just thought I'd give you a heads-up in case you weren't aware.

svaarala commented:
Thanks, I wasn't aware: all of them seem to be:

    duk_error_macros.c:44:118: warning: declaration of ‘index’ shadows a global declaration [-Wshadow]

I don't get that locally which is probably a GCC version thing (`index` is a function provided by strings.h). It's probably best if I rename all `index` variables to something else.

fatcerberus commented:
I fixed the `require.cache` test.  It was a bug in the test: `pig` exports a string, which is compared by value, not by reference.  All the test cases pass now.

svaarala commented:
Ok, I'll give this a read-through and testing tomorrow. I'll compare against @creationix's proposal too to see if every issue there was addressed (or if not, note it for future reference).

fatcerberus commented:
One thing I'd also like to compare against is the property descriptors used by Node.js for e.g. `require.cache`.  Here everything is fully E/W/C (except the global `require` which is non-enumerable to match other built-ins).

svaarala commented:
What might matter is enumerability, and that matters mostly if there's a chance enumerable properties would somehow accidentally make their way into a for-in loop or JSON serialization. For `require.cache`, for example, that would probably not be a concern. Matching Node.js is still a good idea I agree.

svaarala commented:
For documentation it'd be useful to list the features not supported if that's an easy list to come by. I can think of:

- Loading `.json` and `.node` files (the latter being non-applicable anyway)
- The search path isn't provided out of the box, as it's offloaded to the resolve function
- Other stuff?

README.rst notes:

- Provide an explicit link to the Node.js modules page, maybe identify what version the extra tries to mimic?
- Link to wiki page, http://wiki.duktape.org/HowtoNodejsModules.html (?) for detailed step-by-step getting started stuff (pretty essential from practical experience :-)

Other useful documentation issues as a Wiki pull because it's "how to" stuff: https://github.com/svaarala/duktape-wiki/issues/124.

svaarala commented:
And maybe noting at least the ROM footprint ballpark would be nice - I can check that when I test this out.

svaarala commented:
Couple of dist-related issues:

- The extra should be added to `util/make_dist.py`
- The extras/module-node/Makefile test should except paths available in the dist package (it now assumes it's being run in the Duktape repo); pretty arbitrary, but that's how the test Makefiles in other extras are set up

svaarala commented on extras/module-node/README.rst:
```diff
@@ -0,0 +1,62 @@
+      duk_module_node_init(ctx, resolve_cb, load_cb);
+
+* The resolve callback looks like this::
+
```
Here it might be good to say "The resolve callback is a Duktape/C function which [... does what ...], for example:" or something along these lines: it's good to say the function is a standard Duktape/C function so that the argument/return value semantics are clear, and summarizing the purpose ("... resolves a requested ID into a canonical module ID ...") is nice for the reader.

svaarala commented on extras/module-node/README.rst:
```diff
@@ -0,0 +1,62 @@
+  parent ID will be an empty string.
+
+* The load callback looks like this::
+
```
Same here, mention that it's a Duktape/C function.

Maybe mention somewhere that both resolve and load can be Ecmascript functions too?

svaarala commented on extras/module-node/README.rst:
```diff
@@ -0,0 +1,62 @@
+      /* After initializing the Duktape heap or when creating a new
+       * thread with a new global environment:
+       */
+      duk_module_node_init(ctx, resolve_cb, load_cb);
```
If this convention was changed so that `duk_module_node_init()` would expect that callbacks in the value stack, it would then support Ecmascript resolve/load functions. If you want to make versioning easier, the value stack argument could also be an initialization parameter object providing the callbacks which would then be easier to extend:

    {
        "resolve": ...,
        "load": ...
    }

fatcerberus replied:
Assuming that either (or both) of #575 or the format string-based `duk_put_props()` will eventually make it into master, then I'm fine with using an object.

fatcerberus commented:
Technically this could be used to load `.json` files since the load callback doesn't need to return source code - it can assign to `module.exports` itself.

As for the Makefile, I originally copied the Makefile from another extra and actually assumed it was broken because it didn't work from the repo - I thought the repo and dist used to same relative paths--forgot that you can't "make dist" with the dist. :)

svaarala commented:
Ok, for `.json` files it's maybe a topic to cover in the wiki page: it's not directly supported because of the abstraction but you can implement it within the load callback.

Some extras used to have paths like `../../dist/src/` but I think I've fixed them all in master to be compatible with the dist layout instead.

fatcerberus commented:
Yeah, I'll fix the path, and also see about adding to `make_dist.py`.  No promises on that latter one, Python isn't really my forte. :)

svaarala commented:
Regarding the init argument, I can symphatize with trying to avoid the rather verbose object parameter - on the other hand, it makes versioning much easier to deal with, and a fully featured loader like this will quite likely need additional flags or parameters going forward. It'd be annoying if that also meant a change in the value stack behavior or C signature - both violating source compatibility which would then potentially prevent making changes for a year or so, or would necessitate introducing multiple C initializer functions which then need deprecating etc. The cost of a verbose init sequence is far smaller in my view :)

Most extras don't do this for the inits now which is fine if there are no foreseeable changes needing extra parameters. I did add a placeholder flags field to the C signature of most init functions just to ensure there'd be no versioning issues in adding some trivial behavioral flags (for example, one may want to control flushing behavior of the logging framework or console.log writes). This isn't of course a perfect solution because the unforeseen thing may be a value or a function, etc. But versioning is more a matter of avoiding most pitfalls to minimize effort rather than being 100% right.

svaarala commented:
If you want I can make the dist changes too.

svaarala commented:
Also a releases entry is needed, I can add that separately if you wish.

fatcerberus commented:
While I find it annoying in general to work with objects on the value stack, in this case I'm okay with it because it only needs to be done once.

I think my main issue with how things are right now is that it's difficult when scanning value stack code to tell if you've balanced the stack properly.  Verifying correctness often requires reading over the code 4-5 times (this was the case for me with `duk__eval_module_source()` in this extra).  For Duktape/C functions it's not as important since call handling will pop any garbage values for you, but elsewhere you have to be very careful to make sure you balance the stack properly--which often requires keeping the Duktape API documentation open at the same time to check stack signatures.

If the situation is improved so that correctness is clear just from reading the code without needing to cross reference, I'd be much more receptive to Duktape APIs accepting objects on the stack.

fatcerberus commented:
Should `duk_module_node_init()` throw if the object doesn't have "resolve" and "load" properties (or they are not functions)?

svaarala commented:
That'd be my expectation - the other alternative would be for the initialization to finish but remain in a broken state which would be quite unintuitive. Also returning an error code basically replicates the error mechanism (and the caller will still need a protected call to catch any runtime errors which may always happen).

fatcerberus commented:
For future work it might good to eventually support `Duktape.modLoad`/`.modResolve` when using this extra.

Let me know what you think now.

svaarala commented:
Hmm, what do you mean with `Duktape.modLoad` and `Duktape.modResolve`? Looking up the callbacks from there?

I'd be fine with the resolve/load functions to be initialized from C code so that they're by default not exposed to Ecmascript code. There's a native init call site anyway. What I would like, though, is for it to be OK for them to Ecmascript functions and this should now be the case.

Switching the functions after initialization might be necessary in some situations and I guess in that case having them be addressable from Ecmascript code might be useful (but also a sandboxing concern). That case would now require a C helper, and there's no defined way of replacing the functions now; user code would need to interact with the global stash directly.

That's actually another design alternative: the module loader could just assume you register the load/resolve callbacks into the global stash using some documented name. They could then be changed at any time without any need for additional APIs provided by the module loader.

svaarala commented:
That design alternative is basically what Duktape.modSearch() tries to do; the only difference would be that the properties would be in the stash, i.e. hidden by default from Ecmascript code.

svaarala commented on RELEASES.rst:
```diff
@@ -1669,6 +1669,9 @@ Planned
 
+* Add an extra module (extras/module-node) providing a Node.js-like module
+  loading framework supporting `require.cache`, `module.loaded`, etc.
+
```
The formatting is incorrect for RST, should be double backticks for preformatted/code. Though I've mostly avoided formatting in the RELEASES file so that it reads better as plain text.

fatcerberus replied:
Alright, fixed locally. :)

svaarala commented:
The releases file and README.rst both use single backticks for code/preformatted; for RST two backticks are needed.

fatcerberus commented:
Hm, interesting thought on looking up the functions from the stash.  I guess I'll leave it as-is for now and if you decide you want to implement that later you can.

By the way is `Duktape.modSearch` an accessor?  I thought I remember that there was no global `require` function until I provided a module callback.

svaarala commented:
The global `require()` was always present in Duktape 1.x and `Duktape.modSearch` is just a normal property that you assign to. You could also define it as an accessor, though I'm not sure what that would accomplish.

svaarala commented:
Hmm, so with no action there's no easy way to replace the resolve/load callbacks at runtime I guess? That might seem odd, but I've more than once worked with a setup where an initial loader is used to pull in an actual module loader, which then loads the remaining modules.

svaarala commented:
I'm fine with leaving things as they're now, but it's maybe worth having a pull request / issue for Duktape 2.x release to figure out if minor changes are needed to round out the functionality.

fatcerberus commented:
Huh, that's weird.  When I was originally dabbling with the module system I remembered `require` not being defined until I provided a modSearch callback.  Oh well, it doesn't really matter and is unrelated to this pull, just a curiosity.

Anyway yeah, it's technically possible to replace the callbacks later on by replacing `_modLoad` and `_modResolve` in the global stash, but that's kind of "undocumented". :)

svaarala commented:
Ok, well it might be enough to just document that convention for now.

But then, with that documented, having the init function set them for you is somewhat unnecessary: the caller could just set them before (or after) initializing the module framework.

fatcerberus commented:
This might be a worthwhile direction in general, using the stash to share data with Duktape internals in the same way the value stack is used for transient stuff.  We'd need to come up with a formal convention though, to ensure application code using the stash doesn't accidentally step on the Duktape values.

svaarala commented:
Agree, though it probably doesn't need to be very complicated. For example, Duktape internals could just be internal keys on the global stash (`\xffFoobar`) which are anyway reserved for Duktape. The module loader here is not part of internals so it'd need to "compete" with other users, but I don't think that's much of a concrete problem in practice, as long as expressive names are used.

fatcerberus commented:
This is what I really like about Duktape by the way, Ecmascript and C code are generally interchangeable.

So changes to make before merge are:
* Document that callbacks can be swapped out by modifying the relevant properties in the global stash
* Avoid setting callbacks in `duk_module_node_init()`
* Rename stash keys for callbacks (e.g. "moduleLoad"), avoid internal keys

Anything else?

svaarala commented:
I'll re-read the current state in around 3-4 hours so maybe something comes to mind. But this is already in good enough shape for master I think. It's still useful to try to figure out how to minimize mechanisms and/or increase flexibility, etc :)

fatcerberus commented:
By the way it's included in `util\make_dist.py` now.

fatcerberus commented:
By the way, I don't know which Node.js version to say this is based on - their module API has been documented as "Locked" for quite some time. :)

svaarala commented:
Ok, fair enough :)

svaarala commented:
What happens if you re-initialize the module loader for the same global environment? Is this allowed; if so, what's the expected outcome. For example, are module caches reset.

Another question that come to mind is whether the extra works in Duktape 1.x. It's quite clear extras should work with the version they ship with, but there's no reason why they couldn't support other versions too.

The reason I ask is that I think there are Duktape 1.x users who might want to use this loader before migrating to Duktape 2.x.

fatcerberus commented:
This is what happens if you try to double initialize:
* A new global `require` is registered
* The resolve and load callbacks are changed to whatever you pass in
* The module cache is replaced with a fresh object.

That last thing makes it not so harmless: Modules which are already loaded will then have a stale `require.cache`.

And theoretically this should work just fine with Duktape 1.x (right now at least): I don't believe it calls any new/updated APIs. I guess this would be a candidate for merging into the 1.x maintenance branch if there is ever a Duktape 1.6.

svaarala commented:
Hmm, so should we just say "don't initialize twice" for a global environment and avoid providing any guarantees if that's done? If reinitialization is allowed, then the semantics matter and become a versioning concern.

Re: Duktape 1.6, I think it's quite possible such a release will happen as a final release in the 1.x branch with a minimal set of features improving usability for 1.x users (migration takes time and there's often no pressing need to do so).

So I guess we'd then try to keep this extra Duktape 1.x compatible, #ifdef'ing with DUK_VERSION if necessary. Hopefully the extra itself didn't need to be branched.

fatcerberus commented:
I agree, I'll add a note to `README.rst` to not initialize more than once for the same global environment.  Right now the only reason to do so is to replace the callbacks--and that can be done by modifying the stashed properties directly.

fatcerberus commented:
Kind of nitpicky but `duk_module_node_init` seems awkward - should we rename it to `duk_module_init_node()` instead?  I think it would read back better - i.e. "Initialize the module system using Node-like semantics"

svaarala commented:
That seems a bit weird to me, `duk_init_module_node()` would read more like a sentence and would keep the extra name intact (`module_node)`. The other extras use that `duk_xxx_init()` convention to ensure the module name is intact; they should be changed to match if we use a different convention here.

The C module convention uses `dukopen_foo_bar()`. I originally thought of using that convention, but because the initializers make changes to the global object etc (rather than just returning a value) that wouldn't be strictly speaking correct if they were considered modules.

Yet another option would be to model all of the extras as C modules matching the c-module-convention.rst document. In that case the `dukopen_foo_bar()` call would return an object containing methods and constants. You'd then call, say, `init()` from those, as if you'd `require('foo_bar').init()`.

But I suggest we revisit the naming and init conventions separately, as they should ideally be same or similar across the extras.

fatcerberus commented:
@saghul If you want to use this in sjs now, go right ahead - it's basically finished, we're just trying to hash out the little nitpicky details now. :)

svaarala commented:
@fatcerberus Let me know when you've pushed final changes for review :)

fatcerberus commented:
Hehe, sorry, I'm a perfectionist.

It should be good to go now.  I updated the test program to work with the new value stack semantics of `duk_module_node_init()` and everything still checks out.  No further changes should be needed.

saghul commented:
@fatcerberus Thanks mate! I'm following the discussion from the sidelines :-) I have a couple of things to iron out first, then I'll have a deep loop and integrate it.

svaarala commented:
@fatcerberus I'll float the idea of removing the resolve/load registration in the init function once more with a concrete example (for completeness if nothing else :-).

So this is the current init call site:
```
duk_push_object(ctx);
duk_push_c_function(ctx, cb_resolve_module, DUK_VARARGS);
duk_put_prop_string(ctx, -2, "resolve");
duk_push_c_function(ctx, cb_load_module, DUK_VARARGS);
duk_put_prop_string(ctx, -2, "load");
duk_module_node_init(ctx);
```

If the init function didn't set resolve/load but left the user to do that (which the calling code needs to do anyway if the functions are replaced), it would look like:
```
duk_push_global_stash(ctx);
duk_push_c_function(ctx, cb_resolve_module, DUK_VARARGS);
duk_put_prop_string(ctx, -2, "\xffmodResolve");
duk_push_c_function(ctx, cb_load_module, DUK_VARARGS);
duk_put_prop_string(ctx, -2, "\xffmodLoad");
duk_pop(ctx);
duk_module_node_init(ctx);
```

The difference isn't big and strips the init function of handling the resolve/load registration. But maybe we'll leave the init as it exists now, because:

- The current call site will look better with API primitives for multiple property set/get.
- There may be a need for other options besides the load/resolve functions so init arguments / options object will be needed anyway.

fatcerberus commented:
One thing I don't like about setting internal properties directly is that it's prone to error - if the name used starts with A-F the first character(s) will be interpreted as part of the character code (e.g. `"\xfffoo"` gets interpreted as /xfff + "oo").  I ran into that problem a few times in minisphere, so I ended up developing a convention to avoid it: `"\xff" "foo"`"

svaarala commented:
That's actually mentioned in the Guide too:

```
/* Create an internal string, which can then be used to read/write internal
 * properties, and can be passed on to Ecmascript code like any other string.
 * Terminating a string literal after a hex escape is safest to avoid some
 * ambiguous cases like "\xffab".
 */
duk_push_string(ctx, "\xff" "myprop");
```

svaarala commented on extras/module-node/README.rst:
```diff
@@ -0,0 +1,76 @@
+  environment.  Doing so will put the module system in an inconsistent state.
+
+* The resolve callback is a Duktape/C function which takes the string passed
+  to `require()` and resolves it to a canonical module ID (for Node.js this is
```
Markdown quoting for require().

fatcerberus replied:
Force of habit. :-/

svaarala replied:
Yup, same here: it's somewhat annoying to have both Markdown and RST documents in the repo. I started with RST so I've only used Markdown for a few documents people usually expect to be Markdown (like README.md). (Though, Github does support RST for README, CONTRIBUTING, etc too.)

svaarala commented:
Hmm, it's probably best if I fix any of the tiny nits like RST markdown issues in a separate pull so we'll get the main functionality in master. I'll re-read the code and see if there's anything more substantial.

svaarala commented on extras/module-node/README.rst:
```diff
@@ -0,0 +1,76 @@
+
+* The load callback is a Duktape/C function which takes the resolved module ID
+  and either returns the Ecmascript source code for the module, or populates
+  ``module.exports`` itself and returns undefined (useful for C modules)::
```
Just to be sure: it's also allowed to populate symbols into module.exports and *also* return a string (same as with Duktape 1.x)?

Also: I'm assuming it's allowed to replace `module.exports` in the load() function (as with the updated Duktape 1.x behavior)?

fatcerberus replied:
Yes, both of those should work.  The `module` on the stack is the same one passed to the module to be loaded.

fatcerberus replied:
To clarify on the above point: `module.exports` isn't read until just before `duk__handle_require()` returns.  So even if `module.exports` is replaced wholesale by the load callback, `require()` handling will see the new value.

svaarala commented on extras/module-node/test.c:
```diff
@@ -0,0 +1,69 @@
+		"if (module.id == 'pig.js') module.exports = 'you\\'re about to get eaten by %s';"
+		" else if (module.id == 'cow.js') module.exports = require('pig');"
+		" else if (module.id == 'ape.js') module.exports = { module: module, __filename: __filename, wasLoaded: module.loaded, size: 0 };",
+		module_id);
```
This has no functional impact, but wouldn't it be cleaner if `cb_load_module` contained an if-else ladder for the virtual filenames (in C) and would return the module sources for each? Now this returns a fixed string, regardless of module ID, which then exports different symbols.

fatcerberus replied:
That would probably be better, I agree.  This was kind of thrown together in 5 minutes, can you tell? :)

svaarala replied:
It's test code so it doesn't really matter other than if anyone looks, it might give the wrong idea of how to load built-in fixed strings :)

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,263 @@
+
+	/* [ ... stash ... resolved_id ] */
+
+	id = duk_require_string(ctx, -1);
```
A small nit here, can be fixed here or in a follow-up cleanup: because duk_require_string() is used, the duk_is_string() check directly above is not necessary. The error message will be a bit vague but I don't think it needs to be very explicit if the resolve() function is violating the basic contract? If the error stays in place, then the error message ("error in module resolution") should probably be made more explicit because it's actually quite vague now compared to the check which triggers it?

fatcerberus replied:
I agree, the error message is too vague.  I removed the check.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,263 @@
+
+	id = duk_require_string(ctx, -1);
+	if (duk__get_cached_module(ctx, id)) {
+		goto have_module;  /* use the cached module */
```
Goto here is unnecessary I think, because the target will be right after the else-block.

OTOH if the got stays, then the else-block could be de-indented by placing it at the top level.

fatcerberus replied:
I left the goto because it makes it explicit that we're using the cached module.  I did de-indent the else block though, like you said.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,263 @@
+			duk_throw(ctx);  /* rethrow */
+		}
+
+		/* [ ... module source? ] */
```
Very small nit re: comment format: I've used "source?" to indicate that an element might be present or *absent*. Here it means there's either a source string or undefined, so maybe "source/undefined"?

fatcerberus replied:
In the early drafts for this pull (before the callbacks were converted to Duktape/C calls), that actually was the case, if the callback didn't return source there wouldn't be anything on the top of the stack except for the `module` object.  I'll fix the comment.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,263 @@
+
+		/* [ ... module source? ] */
+
+		if (!duk_is_undefined(ctx, -1) && !duk_is_string(ctx, -1)) {
```
This would have smaller footprint:

    if (!duk_check_type_mask(ctx, -1, DUK_TYPE_MASK_UNDEFINED | DUK_TYPE_MASK_STRING)) {
        ...
    }

svaarala replied:
Another structure option would be:

    if (duk_is_string(ctx, -1)) {
        ...
    } else if (duk_is_undefined(ctx, -1) {
        /* nop */
    } else {
        error
    }

fatcerberus replied:
Hm, I forgot that `duk_check_type_mask()` was a thing.

svaarala replied:
That three-part if-statement structure would probably be the cleanest here though.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,263 @@
+	duk_push_string(ctx, "require");
+	duk_def_prop(ctx, -3, DUK_DEFPROP_HAVE_VALUE);
+	duk_push_string(ctx, id);
+	duk_put_prop_string(ctx, -2, "\xffmoduleID");
```
Capitalization trivia: would `\xffmoduleId` be more consistent with using `id` elsewhere?

fatcerberus replied:
My usual convention when using camelCase is to capitalize acronyms and initialisms unless it's the first term in the identifier.  So you'd have `idOfModule` but `moduleID`.  I don't know if there's an established convention for this in Duktape though, I don't object to changing it if so. :)

svaarala replied:
There's not really a very established convention for that but I've tended to capitalize only the first letter of an acronym (`requireId`). Similarly I'd probably favor `encodeJx()` over `encodeJX()`. It doesn't really matter though because it's entirely internal to the module loader.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,263 @@
+	/* Wrap the module code in a function.  This is the easiest way to
+	 * implement CommonJS closure semantics and matches the behavior of e.g.
+	 * Node.js.
+	 */
```
Maybe comment here that `__dirname` is not provided (it's the one real difference in the wrapping). Would it make sense for `__dirname` to exist but be `undefined`?

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,263 @@
+	/*
+	 *  Stack: [ ... module source ] => [ ... module ] (retval == 0)
+	 *         [ ... module source ] => [ ... err ]    (retval != 0)
+	 */
```
This is a bit misleading because errors may be thrown and are not caught here (i.e. `duk_call()` is used to execute the wrapped module).

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,263 @@
+		duk_remove(ctx, -2);  /* [ ... err ] */
+		goto fail;
+	}
+	duk_call(ctx, 0);
```
This may error specifically.

fatcerberus replied:
I didn't think there was anyway the duk_call could error?  It's the equivalent of `eval("(function() { ... })")`.  If it compiles successfully (and the compile *is* protected) I don't think it can throw?

svaarala replied:
Right, the call is evaluating the function expression. It could run out of memory when creating the function instance but that wouldn't be an "expected error" like something being wrong with the module.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,263 @@
+		}
+
+		if (duk_is_string(ctx, -1)) {  /* if we have source */
+			if (duk__eval_module_source(ctx) != DUK_EXEC_SUCCESS) {
```
One approach here would be to allow `duk__eval_module_source()` to freely throw errors (it also needs to return code in that case) but use `duk_safe_call()` here to catch the errors more reliably.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,263 @@
+
+	duk_idx_t options_idx;
+
+	duk_require_object_coercible(ctx, -1);
```
This is perhaps an unnecessary check. If the argument is not object coercible (null or undefined) the first property read will throw anyway.

svaarala replied:
Hmm, well, `\xffrequireCache` gets written before that so maybe this is useful after all.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,263 @@
+	duk_put_prop_string(ctx, -2, "\xffrequireCache");
+	duk_pop(ctx);
+
+	/* stash callbacks for later use */
```
Maybe worth commenting here that user code can later modify the callbacks by overwriting them from the global stash.

svaarala commented:
Ok, I've re-read the code. Trivia stuff is in the commit comments, I can fix them in a cleanup pull if you prefer.

Two maybe not-so-trivia comments:

- Should `__dirname` be provided by the function wrapper with value set to `undefined`? If this is not done and a module reads `__dirname` it will lookup a global which doesn't exist and will trigger a `ReferenceError` which may not be the best behavior.
- An alternative to structuring the call to `duk__eval_module_source()` would be to allow `duk__eval_module_source()` to freely throw on errors and use a `duk_safe_call()` to call it. As things are now, there's at least one unprotected `duk_call()` in `duk__eval_module_source()` and potential for a few of the other operations to also throw.

fatcerberus commented:
I think I agree re: `__dirname` being present but undefined.  Getting an error thrown on attempting to access it (in strict mode) would be a nasty surprise and hurts Node.js compatibility.  It's trivial to add an argument to the wrapper, I'll add it.

fatcerberus commented:
w.r.t. `duk__eval_module_source()` throwing: Overall I agree it'd be cleanest to just wrap it in a safe call and let it go wild.  That'd probably help footprint too since there would be fewer explicit checks needed.

svaarala commented:
@fatcerberus I've got to sleep but I'll pick up tomorrow. Let me know if you're making any fixes - if not, I'll merge and open a trivia cleanup pull. :)

fatcerberus commented:
I need to update `duk__eval_module_source()` anyway so I'll address the trivia too tonight.  Then you can give it a final review tomorrow and merge to master.

fatcerberus commented:
@svaarala Alright, this is merge-ready again.  `duk__eval_module_source()` is now a safe call.

svaarala commented on extras/module-node/duk_module_node.c:
```diff
@@ -0,0 +1,275 @@
+	}
+
+	if (duk_is_string(ctx, -1)) {
+		duk_int ret;
```
Typo in ret typing, `duk_int_t`.

svaarala commented:
There are some small bugs in the pull (e.g. `duk_int` instead of `duk_int_t`) but I'll merge and open a pull for the fixes because they're so small.

fatcerberus commented:
Sorry, I don't get up early enough or I would have fixed that `duk_int` bug before you merged.

svaarala commented:
No worries, I wanted this in master and I had already bugged you with trivia quite a bit :)

fatcerberus commented:
By the way I tested - this compiles successfully against Duktape 1.x :-)
